### PR TITLE
Add locale provider

### PR DESCRIPTION
**Pull request by acostalima**
## TO DO

- [x] Check whether Android runtime needs to be polyfilled
- [x] Rename provider and export to `LocaleProvider`
- [x] Rename consumer and export to `LocaleConsumer`
- [x] Rename hook and export to `useLocale`
- [x] Rename HOC and export to `withLocale`

### CODE REVIEW

File: android/app/build.gradle [hunk 1]
```diff
@@ -121,7 +121,8 @@ def enableProguardInReleaseBuilds = false
  * give correct results when using with locales other than en-US.  Note that
  * this variant is about 6MiB larger per architecture than default.
  */
-def jscFlavor = 'org.webkit:android-jsc:+'
+// def jscFlavor = 'org.webkit:android-jsc:+'
+def jscFlavor = 'org.webkit:android-jsc-intl:+'
```
Comment (acostalima) [inline android/app/build.gradle:125]: This statement is enough to polyfill the runtime!

File: src/shared/modules/react-native-locale/ReactNativeLocaleProvider.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        const nextLocale = findLocaleById(locales, localeId);
+
+        if (!nextLocale) {
+            return;
+        }
+
+        setCurrentLocale(nextLocale);
+
+        if (localeId !== currentLocale.id) {
+            onLocaleChange?.(nextLocale.id);
```
Comment (ruipneves) [inline src/shared/modules/react-native-locale/ReactNativeLocaleProvider.js:49]: We will maintain this? 😛

Comment (acostalima) [reply src/shared/modules/react-native-locale/ReactNativeLocaleProvider.js:49]: Well, no harm's done if we keep the callback as optional. I'd rather offer the possibility of attaching a callback now than later realize we need it for some reason and not having it in place.

Comment (ruipneves) [reply src/shared/modules/react-native-locale/ReactNativeLocaleProvider.js:49]: I do not like having a provider cause other unexpected effects from a callback sent to him. To me whoever consumes it should receive the `locale` and a `setLocale` and when the locale changes or whenever the developer wants to modify the locale he should do it in the `useCallback` where he calls `setLocale`. But the rest can give their opinion

Comment (acostalima) [reply src/shared/modules/react-native-locale/ReactNativeLocaleProvider.js:49]: > I do not like having a provider cause other unexpected effects from a callback sent to him. To me whoever consumes it should receive the `locale` and a `setLocale` and when the locale changes or whenever the developer wants to modify the locale he should do it in the `useCallback` where he calls `setLocale`.

That's exactly what I had in mind. If the possibility to change the language in-app is offered by any given app, the approach one should take is the one you've mentioned. However, that only works if we're considering only React components listen to locale changes. What if we come across a scenario where a module is interested in locale changes as well?

Comment (ruipneves) [reply src/shared/modules/react-native-locale/ReactNativeLocaleProvider.js:49]: Then you would just need to use create a hook with a `useEffect` with the `locale` as a dependency and import the module and have it return whatever the module returns 😛 less magic, more coherent in my opinion.
@satazor @fsdiogo what do you think?

Comment (acostalima) [reply src/shared/modules/react-native-locale/ReactNativeLocaleProvider.js:49]: > Then you would just need to create a hook with a `useEffect` with the `locale` as a dependency and import the module and have it return whatever the module returns 😛 less magic, more coherent in my opinion.
> @satazor @fsdiogo what do you think?

If the module is not a React component nor does have dependencies on React, it really doesn't make much sense to create a React component with a `useEffect` just for that purpose. Also, that component would not render anything at all which makes it kind of useless. 😕

**Pull request merged**